src/main/java/picard/sam/SAMSplitter.java

```diff
@@ -0,0 +1,183 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package picard.sam;
+
+import htsjdk.samtools.*;
+import htsjdk.samtools.util.IOUtil;
+import htsjdk.samtools.util.Log;
+import htsjdk.samtools.util.ProgressLogger;
+import org.broadinstitute.barclay.argparser.Argument;
+import org.broadinstitute.barclay.argparser.CommandLineProgramProperties;
+import org.broadinstitute.barclay.help.DocumentedFeature;
+import picard.PicardException;
+import picard.cmdline.CommandLineProgram;
+import picard.cmdline.StandardOptionDefinitions;
+import picard.cmdline.programgroups.SamOrBam;
+
+import java.io.File;
+import java.text.DecimalFormat;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * <p/>
+ * Splits the input queryname sorted or querygrouped SAM/BAM file and writes it into
+ * multiple BAM files of approximately equal size. This will retain the sort order
+ * within each output BAM and if the BAMs are concatenated in order (output files are named
+ * numerically) the order of the reads will match the original BAM. As this tool is
+ * intended for unmapped input files, all secondary and supplementary reads will be removed.
+ */
+@CommandLineProgramProperties(
+        summary = SAMSplitter.USAGE_SUMMARY + SAMSplitter.USAGE_DETAILS,
+        oneLineSummary = SAMSplitter.USAGE_SUMMARY,
+        programGroup = SamOrBam.class)
+@DocumentedFeature
+public class SAMSplitter extends CommandLineProgram {
+    static final String USAGE_SUMMARY = "Splits a SAM or BAM file to multiple BAMs.";
+    static final String USAGE_DETAILS = "This tool splits the input SAM/BAM file into multiple BAM files " +
+            "This can be used to split a large unmapped BAM in order to parallelize alignment."+
+            "<br />" +
+            "<h4>Usage example:</h4>" +
+            "<pre>" +
+            "java -jar picard.jar SAMSplitter \\<br />" +
+            "     I=paired_unmapped_input.bam \\<br />" +
+            "     OUTPUT_DIR=out_dir \\ <br />" +
+            "     TOTAL_READS_IN_INPUT=800000000" +
+            "     SPLIT_TO_N_TEMPLATES=24000000" +
+            "</pre>" +
+            "<hr />";
+    @Argument(doc = "Input SAM/BAM file to split", shortName = StandardOptionDefinitions.INPUT_SHORT_NAME)
+    public File INPUT;
+
+    @Argument(shortName = "N", doc = "Split to have approximately N templates per output file.")
+    public int SPLIT_TO_N_TEMPLATES;
+
+    @Argument(shortName = "TOTAL_READS", doc = "Total number of reads in the input file.")
+    public int TOTAL_READS_IN_INPUT;
+
+    @Argument(shortName = "ODIR", doc = "Directory in which to output the split BAM files.")
+    public File OUTPUT_DIR;
+
+    @Argument(shortName = "PAIR", doc = "If true the input bam contains paired reads and " +
+            "each template is assumed to contain two reads (rather than one for unpaired inputs).")
+    public boolean PAIRED_INPUT = true;
```

Comment:
Can you find a way not to have this?  Count by templates? Examine the first few records to see if they are paired end (I'd prefer just to say templates).